src/libutil/file-system.hh

```diff
@@ -165,6 +166,15 @@ inline Paths createDirs(PathView path)
     return createDirs(Path(path));
 }
 
+/**
+ * Set the access and modification times of the given path, not
+ * following symlinks.
+ *
+ * Whether or not the path is a symlink is taken from the mode of `st`
+ * in fallback code.
```

Comment:
I don't think that precision really makes sense without the context of that PR (what's “fallback mode”?)

Besides that, I don't think we should trust `st` to tell us whether the file is a symlink or not. Barring possible race conditions (which I think we can't avoid any way), there's no guaranty that `st` refers to the path at `path` (might have been a totally made up struct just to set the time, or taken from another file). We should rather re-check within the function ([std::filesystem::is_symlink](https://en.cppreference.com/w/cpp/filesystem/is_symlink) being our friend here I guess)

Reply:
fallback mode is without `lutimes`, with just `utimes` (yes, this should be reworded)

> Besides that, I don't think we should trust `st` to tell us whether the file is a symlink or not. 

Note that this is not a new thing. I agree in principle, but how would you update `canonicaliseTimestampAndPermissions` in practice? It is, already, very geared about working from a prior stat (in a context that we don't think things would change without them looking.

Unless we want to scope creep and overhaul `canonicaliseTimestampAndPermissions`, I think it would be good to just document the race condition on this function and say callers must be confident `st` is up to date.

Reply:
OK this is done now. the `struct stat` way still exists, but is not forced upon eveyone else.